test/distributed/test_c10d_pypg.py

```diff
@@ -0,0 +1,154 @@
+# Owner(s): ["oncall: distributed"]
+
+import os
+
+import torch
+import torch.distributed as dist
+from torch.testing._internal.common_utils import (
+    run_tests,
+)
+from torch.futures import Future
+import torch.nn as nn
+from torch.nn.parallel import DistributedDataParallel as DDP
+import test_c10d_common
+import weakref
+from torch._C._distributed_c10d import _create_work_from_future
+from torch.testing._internal.common_distributed import (
+    MultiProcessTestCase,
+)
+
+def create_work(result):
```

Comment:
Crazy question: could we just enforce that Python process groups return futures from their collectives? cc @mrshenli

Reply:
I looked into it, it's not a lot of work to replace work with future, but there are a few features we'd lose that we should first assess for usefulness:

- timed wait - easy to implement and probably useful
- cancelation - only implemented for GLOO send/recv
- source rank when doing `dist::recv` from any rank - seems pretty important
- Future::wait() with NCCL when NCCL_BLOCKING_WAIT is set - Work::wait blocks, Future::wait doesn't.